components/function-controller/pkg/controller/function/function_controller.go

```diff
@@ -107,308 +100,360 @@ var (
-	if err != nil && errors.IsNotFound(err) {
-		log.Info("Creating the Function's ConfigMap", "namespace", deployCm.Namespace, "name", deployCm.Name)
-		err = r.Create(context.TODO(), deployCm)
-		if errors.IsNotFound(err) {
-			return reconcile.Result{}, nil
+	return r.updateFunctionConfigMap(currentCm, desiredCm)
+}
+
+// getOrCreateFunctionConfigMap returns the existing Function ConfigMap or
+// creates it from the given desired state if it does not exist.
+func (r *ReconcileFunction) getOrCreateFunctionConfigMap(desiredCm *corev1.ConfigMap) (*corev1.ConfigMap, error) {
+	ctx := context.TODO()
+
+	currentCm := &corev1.ConfigMap{}
+	err := r.Get(ctx,
+		client.ObjectKey{
+			Name:      desiredCm.Name,
+			Namespace: desiredCm.Namespace,
+		},
+		currentCm,
+	)
+
+	switch {
+	case errors.IsNotFound(err):
+		// TODO(antoineco): a Kubernetes event would be more suitable than a log entry
+		log.Info("Creating Function ConfigMap", "namespace", desiredCm.Namespace, "name", desiredCm.Name)
+
+		if err := r.Create(ctx, desiredCm); err != nil {
+			return nil, err
 		}
-		return reconcile.Result{}, err
+		return desiredCm, nil
 
-	} else if err != nil {
-		return reconcile.Result{}, err
+	case err != nil:
+		return nil, err
 	}
 
-	log.Info("Function ConfigMap:", "namespace", deployCm.Namespace, "name", deployCm.Name)
-
-	return reconcile.Result{}, nil
+	return currentCm, nil
 }
 
-// Update found Function's ConfigMap
-func (r *ReconcileFunction) updateFunctionConfigMap(foundCm *corev1.ConfigMap, deployCm *corev1.ConfigMap) error {
+// updateFunctionConfigMap reconciles the current Function ConfigMap with its
+// desired state.
+func (r *ReconcileFunction) updateFunctionConfigMap(currentCm, desiredCm *corev1.ConfigMap) (*corev1.ConfigMap, error) {
+	if reflect.DeepEqual(desiredCm.Data, currentCm.Data) {
+		return currentCm, nil
+	}
 
-	if !reflect.DeepEqual(deployCm.Data, foundCm.Data) {
-		foundCm.Data = deployCm.Data
-		foundCm.TypeMeta = deployCm.TypeMeta
-		foundCm.ObjectMeta = deployCm.ObjectMeta
-		log.Info("Updating Function's ConfigMap", "namespace", deployCm.Namespace, "name", deployCm.Name)
-		err := r.Update(context.TODO(), foundCm)
-		if err != nil && !errors.IsNotFound(err) {
-			return err
-		}
+	// TODO(antoineco): a Kubernetes event would be more suitable than a log entry
+	log.Info("Updating Function ConfigMap", "namespace", desiredCm.Namespace, "name", desiredCm.Name)
 
+	newCm := &corev1.ConfigMap{
+		ObjectMeta: desiredCm.ObjectMeta,
+		Data:       desiredCm.Data,
 	}
+	newCm.ResourceVersion = currentCm.ResourceVersion
 
-	return nil
-
+	if err := r.Update(context.TODO(), newCm); err != nil {
+		return nil, err
+	}
+	return newCm, nil
 }
 
-func (r *ReconcileFunction) getFunctionBuildTemplate(fn *serverlessv1alpha1.Function, ri *runtimeUtil.RuntimeInfo) error {
+// generateFunctionHash generates a hash from the Function config.
+func generateFunctionHash(fnCm *corev1.ConfigMap) (string, error) {
```

Comment:
Should we have a test for this?

Reply:
As an improvement, but not as part of the refactoring.